File: C++/Binary Search/Right_interval.cpp
// Leetcode - medium
// Sort + Binary Search - O(NlogN) 
// Author: Pavan Kumar Paluri

class Solution {
public:
    int bs(vector<vector<int>>& in, int v) {
        int l=0,n=in.size(),h=n-1;
        
        while(l<=h) {
            int m = l + (h-l)/2;
            
            if(in[m][0] < v) {
                l=m+1;
            }
            else {
                h=m-1;
            }
        }
        
        return l<n ? l : -1;
    }
    
    
    vector<int> findRightInterval(vector<vector<int>>& in) {
        
        int n = in.size();
        vector<int> ans (n, -1); 
        
	// Store the index of each interval along side
	// the start and end points.
	// Will help post-sorting.
        for(int i=0; i<n; ++i) {
            in[i].push_back(i);
        }
        
	// Sort in non-decreasing order with respect to the start points.
	// If start points are equal,  use end points.
        sort(in.begin(), in.end(), [](const vector<int>& a, const vector<int>& b) {
            if(a[0] == b[0])
                return a[1]<b[1];
            return a[0]<b[0];
        });
        
        for(int i=0; i<n; ++i) {
            int ix = in[i][2];
            int px = bs(in, in[i][1]);
            if(px != -1)
                ans[ix] = in[px][2];
        }
        
        return ans;
    }
};

File: Python/Binary Search/Right_interval.py
# Find the minimum right interval 
# Using Sorting + Binary Search in O(NlogN) time complexity
# Author: Pavan Kumar Paluri
# Problem: https://leetcode.com/problems/find-right-interval/

class Solution:
    def findRightInterval(self, intervals: List[List[int]]) -> List[int]:
        # append index to the list 
        for (x,y) in enumerate(intervals):
            intervals[x].append(x)
        print(intervals)
        # sort the intervals based on x-value, if tie: based on second value
        output = [-1]*len(intervals)
        intervals = sorted(intervals, key=lambda a:a[0])
        hi = len(intervals)-1
        low =0
        def binary_search(intervals, low, hi, target):
            while low <= hi:
                mid = low + (hi-low)//2
                if intervals[mid][0] < target:
                    low = mid+1
                else:
                    # intended target val is > mid
                    hi= mid-1
            return low if low < len(intervals) else -1
        for i in range(len(intervals)):
            val = binary_search(intervals, low, hi, intervals[i][1])
            if val != -1:
                output[intervals[i][2]] = intervals[val][2]
        return output
            
